README.rst
.. |license| image:: https://img.shields.io/github/license/mghorbani2357/mockend
    :target: https://raw.githubusercontent.com/mghorbani2357/mockend/master/LICENSE
    :alt: GitHub Licence

.. |downloadrate| image:: https://img.shields.io/pypi/dm/mockend
    :target: https://pypistats.org/packages/mockend

.. |wheel| image:: https://img.shields.io/pypi/wheel/mockend  
    :target: https://pypi.python.org/pypi/mockend
    :alt: PyPI - Wheel

.. |pypiversion| image:: https://img.shields.io/pypi/v/mockend  
    :target: https://pypi.python.org/pypi/mockend
    :alt: PyPI

.. |format| image:: https://img.shields.io/pypi/format/mockend
    :target: https://pypi.python.org/pypi/mockend
    :alt: PyPI - Format

.. |downloads| image:: https://static.pepy.tech/personalized-badge/mockend?period=total&units=international_system&left_color=grey&right_color=blue&left_text=Downloads
    :target: https://pepy.tech/project/mockend

.. |lastcommit| image:: https://img.shields.io/github/last-commit/mghorbani2357/Mockend 
    :alt: GitHub last commit
    
.. |lastrelease| image:: https://img.shields.io/github/release-date/mghorbani2357/Mockend   
    :alt: GitHub Release Date

.. |codequality| image:: https://app.codacy.com/project/badge/Grade/c1e3c9bb67204f199026f4d6b480a5a9
    :target: https://www.codacy.com/gh/mghorbani2357/Mockend/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=mghorbani2357/Mockend&amp;utm_campaign=Badge_Grade
    :alt: Code quality

.. |codacycoverage| image:: https://app.codacy.com/project/badge/Coverage/c1e3c9bb67204f199026f4d6b480a5a9
    :target: https://www.codacy.com/gh/mghorbani2357/Mockend/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=mghorbani2357/Mockend&amp;utm_campaign=Badge_Coverage
    :alt: Code coverage

.. |workflow| image:: https://img.shields.io/github/workflow/status/mghorbani2357/mockend/main?logo=github
    :alt: GitHub Workflow Status

.. |readthedocs| image:: https://readthedocs.org/projects/mockend/badge/?version=latest
    :target: https://mockend.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

*****************
Mockend API
*****************

.. class:: center

 |license| |workflow| |readthedocs| |codacycoverage| |codequality| |downloadrate| |downloads| |pypiversion| |format| |wheel| |lastcommit| |lastrelease|


Mockend is a Python library that can be used to mock any REST API endpoint.

Installation
============

.. code-block:: bash

    pip install mockend


Quick Start
===========

Mockend is a simple, lightweight, and extensible REST API mocking Python library.
It can be used to mock any REST API endpoint, and can be used to mock any HTTP method.
the library is very easy to use and easy to extend. It just need configuration file, then it
will simulate the REST API response.

How to use
=======================

.. code-block:: bash

    mockend -c config.json

    * Serving Flask app 'mockend.__main__' (lazy loading)
    * Environment: production
    WARNING: This is a development server. Do not use it in a production deployment.
    Use a production WSGI server instead.
    * Debug mode: on
    * Running on http://localhost:5555 (Press CTRL+C to quit)
    * Restarting with stat
    * Debugger is active!
    * Debugger PIN: 141-969-228


In endpoint configuration available method with their configuration can be used to mock any HTTP method.
All configuration of response can be mock in method configuration.

Configuration examples
=======================

.. code-block:: json

    {
        "users": {
            "get": {
                "headers": {...},
                "status_code": "...",
                "mimetype": "...",
                "content_type": "...",
                "direct_passthrough": "...",
                "response": json_data
            }
        }
    }


To define nested endpoint configuration just define the nested endpoint name in the endpoint name.

Nested endpoints
=======================

.. code-block:: json

    {
        "users": {
            "JohnDoe": {
                "get": {...}
            }
        }
    }


To delay the response of endpoint just define delay attribute with a float value.

Delay endpoints
=======================

.. code-block:: json

    {
        "users": {
            "get": {
                "delay": 0.1,
                ...
            }
        }
    }

To generate chuck response set the chuck attribute with a True value and chunk size in chunk_size attribute.

chunk response
=======================

.. code-block:: json

    {
        "users": {
            "get": {
                "chunked": True,
                "chunk_size": 2,
                ...
            }
        }
    }


To simulate dummy to generate same response for same request just define dummy attribute with a True value.

Dummy endpoint
=======================

.. code-block:: json

    {
        "users": {
            "post": {
                "dummy": True
            }
        }
    }

Interactive mode is available to interact with the mockend server.
To start interactive mode just run the following command.
It is easy to use. Just define method and send request to create, update, delete or retrieve data

Interactive mode
=======================

.. code-block:: json

    {
        "users": {
            "interactive": True,
            "get": {...},
            "post": {...}
            "data" : {...}
        }
    }

Documentation: https://mghorbani2357.github.io/Mockend/

HISTORY.rst
.. :changelog:

Changelog
=========

v0.4.0
------
Release Date: 2022-04-29
    * New Features:
        * Chunk response simulator
        * Response abortion
        * Response delay
        * Support SSL
        * Interactive mode
        * Dummy mode

v0.1.1
------
Release Date: 2022-04-21
    * Basic mock api.

tests/mockend.py
from unittest import TestCase


class MockEnd(TestCase):
    def test_none(self):
        pass


tests/test_mockend.py
import json
import time
from copy import deepcopy
from unittest import TestCase
from mockend import config, app, validate_path


class TestEndPoints(TestCase):
    def setUp(self):
        self.app = app.test_client()

    def test_simple_endpoint(self):
        config.update({
            "users": {
                "get": {
                    "response": {
                        "user_ids": [1, 2, 3]
                    }
                }
            }
        })

        self.assertEqual(self.app.get('/users').data, b'{"user_ids": [1, 2, 3]}')

    def test_nested_endpoints(self):
        config.update({
            "users": {
                "JohnDoe": {
                    "get": {
                        "response": {
                            "name": "John Doe",
                            "email": "J.Doe@email.com"
                        }
                    }
                }
            }
        })

        response = json.dumps({
            "name": "John Doe",
            "email": "J.Doe@email.com"
        }).encode('utf-8')

        self.assertEqual(self.app.get('/users/JohnDoe').data, response)

    def test_delay_endpoint(self):
        config.update({
            "users": {
                "get": {
                    "delay": 0.1,
                    "response": {
                        "user_ids": [1, 2, 3]
                    }
                }
            }
        })

        run_time = time.time()

        self.assertEqual(self.app.get('/users').data, b'{"user_ids": [1, 2, 3]}')
        self.assertGreaterEqual(0.1, time.time() - run_time)

    def test_chuck_response(self):
        config.update({
            "users": {
                "get": {
                    "chunked": True,
                    "chunk_size": 2,
                    "response": {
                        "user_ids": [1, 2, 3]
                    }
                }
            }
        })

        self.assertEqual(self.app.get('/users').data, b'{"user_ids": [1, 2, 3]}')

    def test_interactive_mode_pagination(self):
        config.update({
            "users": {
                "interactive": True,
                "pagination": True,
                "get": {},
                "post": {}
            }
        })

        self.app.post('/users/JohnDoe', data=json.dumps(
            {
                "name": "John Doe",
                "email": "J.Don@email.com"
            }
        ).encode('utf-8'))

        self.app.post('/users/Alice', data=json.dumps(
            {
                "name": "Alice",
                "email": "alice@email.com"
            }
        ).encode('utf-8'))

        self.app.post('/users/Bob', data=json.dumps(
            {
                "name": "Bob",
                "email": "bob@email.com"
            }
        ).encode('utf-8'))

        print(self.app.get('/users?start=Alice&limit=2').data)

        self.assertEqual(self.app.get('/users?start=Alice&limit=2').data, json.dumps(
            {
                "Alice":
                    {
                        "name": "Alice",
                        "email": "alice@email.com"
                    },
                "Bob": {
                    "name": "Bob",
                    "email": "bob@email.com"}
            }).encode('utf-8'))

    def test_invalid_path(self):
        self.assertEqual(self.app.get('/users/invalid').status_code, 404)

    def test_invalid_method(self):
        config.update({
            "users": {
                "get": {
                    "response": {
                        "user_ids": [1, 2, 3]
                    }
                }
            }
        })

        self.assertEqual(self.app.post('/users').status_code, 405)

    def test_abortion(self):
        config.update({
            "users": {
                "get": {
                    "abort": 503
                }
            }
        })

        self.assertEqual(self.app.get('/users').status_code, 503)

    def test_dummy_mode(self):
        config.update({
            "users": {
                "post": {
                    "dummy": True
                }
            }
        })

        request_body = json.dumps({
            "user_ids": [1, 2, 3]
        }).encode('utf-8')

        self.assertEqual(self.app.post('/users', data=request_body).data, request_body)

    def test_interactive_invalid_path(self):
        config.update({
            "users": {
                "interactive": True,
                "get": {},
                "post": {}
            }
        })

        self.assertEqual(self.app.get('/users/invalid').status_code, 404)

    def test_interactive_insert(self):
        config.update({
            "users": {
                "interactive": True,
                "get": {},
                "post": {}
            }
        })

        self.app.post('/users/JohnDoe', data=json.dumps(
            {
                "name": "John Doe",
                "email": "J.Don@email.com"
            }
        ).encode('utf-8'))

        self.assertEqual(self.app.get('/users').data, json.dumps(
            {
                "JohnDoe": {
                    "name": "John Doe",
                    "email": "J.Don@email.com"
                }
            }
        ).encode('utf-8'))

        self.assertEqual(self.app.get('/users/JohnDoe').data, json.dumps(
            {

                "name": "John Doe",
                "email": "J.Don@email.com"

            }
        ).encode('utf-8'))

    def test_interactive_update(self):
        config.update({
            "users": {
                "interactive": True,
                "get": {},
                "post": {},
                "put": {},
            }
        })

        self.app.post('/users/JohnDoe', data=json.dumps(
            {
                "name": "John Doe",
                "email": "J.Don@email.com"
            }
        ).encode('utf-8'))

        self.app.put('/users/JohnDoe', data=json.dumps(
            {
                "name": "John Doe",
                "email": "J.Don@gmail.com"
            }
        ).encode('utf-8'))

        self.assertEqual(self.app.get('/users/JohnDoe').data, json.dumps(
            {

                "name": "John Doe",
                "email": "J.Don@gmail.com"

            }
        ).encode('utf-8'))

    def test_interactive_delete(self):
        config.update({
            "users": {
                "interactive": True,
                "get": {},
                "post": {},
                "delete": {},
            }
        })

        self.app.post('/users/JohnDoe', data=json.dumps(
            {
                "name": "John Doe",
                "email": "J.Don@email.com"
            }
        ).encode('utf-8'))

        self.app.delete('/users/JohnDoe')

        response = json.loads(self.app.get('/users').data)

        self.assertNotIn("JohnDoe", response)


class TestPathValidator(TestCase):
    def test_simple_path_validation(self):
        path = "/users/"
        configuration = {
            "users": {
                "get": {}
            }
        }
        paths, path_config, identifier = validate_path(path, configuration)
        self.assertEqual(paths, ["users"])
        self.assertEqual(path_config, configuration["users"])
        self.assertEqual(identifier, None)

    def test_nested_path_validation(self):
        path = "/users/JohnDoe/posts/"
        configuration = {
            "users": {
                "JohnDoe": {
                    "posts": {
                        "get": {}
                    }
                }
            }
        }

        paths, path_config, identifier = validate_path(path, configuration)
        self.assertEqual(paths, ["users", "JohnDoe", "posts"])
        self.assertEqual(path_config, configuration["users"]["JohnDoe"]["posts"])
        self.assertEqual(identifier, None)

    def test_interactive_mode_path_validation(self):
        path = "/users/JohnDoe/"
        configuration = {
            "users": {
                "interactive": True,
            }
        }

        paths, path_config, identifier = validate_path(path, configuration)
        self.assertEqual(paths, ["users", "JohnDoe"])
        self.assertEqual(path_config, configuration["users"])
        self.assertEqual(identifier, "JohnDoe")


mockend/__main__.py
import argparse
import json
from . import app, config

parser = argparse.ArgumentParser(prog='PROG', description='Mockend Service')
parser.add_argument('-c', '--config', metavar='', type=str, required=True, default='config.json', help='Path to the configuration file.')
parser.add_argument('-i', '--host', metavar='', type=str, required=False, default='localhost', help='Host address')
parser.add_argument('-p', '--port', metavar='', type=int, required=False, default=5555, help='Port number')
parser.add_argument('-d', '--debug', metavar='', type=bool, required=False, default=True, help='Debug mode')
parser.add_argument('-e', '--certificate', metavar='', type=str, required=False, default=None, help='Certificate file')
parser.add_argument('-k', '--key', metavar='', type=str, required=False, default=None, help='Key file')
args = parser.parse_args()

config.update(json.load(open(args.config)))
ssl_context = (args.certificate, args.key) if args.certificate and args.key else None
app.run(host=args.host, port=args.port, debug=args.debug, ssl_context=ssl_context)


mockend/__init__.py
"""Mockend API to create fake api endpoints."""

__author__ = 'm.ghorbani2357@gmail.com'

from ._version import get_versions

import json
import time

from flask import Flask, abort, Response, request

app = Flask(__name__)
all_methods = ['GET', 'POST', 'PUT', 'DELETE', 'PATCH']
config = {}


def validate_path(path, configuration):
    """
    Args:
        path(str): incoming request path
        configuration(dict): config dict
    Returns:
        (dict|none): returns config if path is valid, None otherwise

    """
    subpaths = list(filter(''.__ne__, path.split('/')))

    for index, sub_path in enumerate(subpaths):
        if sub_path in configuration.keys():
            configuration = configuration.get(sub_path)
            if configuration.get("interactive", False) and index + 1 == len(subpaths) - 1:
                return subpaths, configuration, subpaths[index + 1]
        else:
            return None, None, None

    return subpaths, configuration, None


def generate_chunk(data, chunk_size):
    """
    Args:
        data(str): incoming request data
        chunk_size(int): chunk size
    Returns:
        (str): returns chunked data
    """
    for i in range(0, len(data), chunk_size):
        yield data[i:i + chunk_size]


@app.route('/', defaults={'path': ''})
@app.route('/<path:path>', methods=all_methods)
def mockend_service(path):
    """
    Args:
        path(str): incoming request path
    Returns:
        (Response): returns flask response
    """
    paths, path_config, identifier = validate_path(path, config)

    if path_config:
        time.sleep(path_config.get('delay', 0))

        if request.method.lower() in path_config:
            method_config = path_config.get(request.method.lower())
        else:
            return abort(405)

        if abortion_code := method_config.get('abort', None):
            return abort(abortion_code)

        if method_config.get("dummy", False):
            return Response(
                response=request.data,
                status=method_config.get("status"),
                headers=request.headers.__dict__,
                content_type=method_config.get("content_type"),
                mimetype=request.mimetype,
                direct_passthrough=method_config.get("direct_passthrough"),
            )

        response_body = method_config.get('response')
        response_body = json.dumps(response_body) if type(response_body) in (dict, list) else response_body
        data = json.loads(request.data) if request.data else {}
        if path_config.get('interactive', False):
            if 'data' not in path_config:
                path_config['data'] = {}
            if request.method.lower() == 'get':
                if paths[-1] == identifier:
                    if identifier not in path_config['data']:
                        abort(404)
                    response_body = json.dumps(path_config['data'].get(identifier))
                else:
                    if path_config.get('pagination', False):
                        ordered_keys = sorted(list(path_config['data'].keys()))
                        pagination_keys = ordered_keys[ordered_keys.index(request.args.get('start')):
                                                       ordered_keys.index(request.args.get('start')) +
                                                       int(request.args.get('limit'))]
                        response_body = json.dumps({key: path_config['data'].get(key) for key in pagination_keys})
                    else:
                        response_body = json.dumps(path_config['data'])

            elif request.method.lower() == 'post':
                if identifier not in path_config.get('data', {}).keys():
                    path_config['data'][identifier] = {}
                path_config['data'][identifier].update(data)

            elif request.method.lower() in ('put', 'patch'):
                path_config['data'][identifier] = data

            elif request.method.lower() == 'delete':
                del path_config['data'][identifier]

        if method_config.get("chunked", False):
            response_body = generate_chunk(response_body, method_config.get("chunk_size", 1))

        return Response(
            response=response_body,
            status=method_config.get("status"),
            headers=method_config.get("headers"),
            mimetype=method_config.get("mimetype"),
            content_type=method_config.get("content_type"),
            direct_passthrough=method_config.get("direct_passthrough"),
        )
    else:
        abort(404)


__version__ = get_versions()['version']
del get_versions
